### PR TITLE
Setup event caching layer for band node

### LINKED ISSUE #827: Setup event caching layer for band node
@taobun already has a POC version of it so let's formalize it into the repo!

Build an abstraction that allows band node to be run while connecting to a SQL database. Logs out when an event is emitted, whether during DeliverTx or EndBlock.

### PR DESCRIPTION
**Pull request by taobun**
fixed #827

### CODE REVIEW

File: chain/db/db.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package db
+
+import (
+	"fmt"
+
+	"github.com/jinzhu/gorm"
+	_ "github.com/jinzhu/gorm/dialects/sqlite"
```
Comment (sorawit) [inline chain/db/db.go:8]: Can you put postgres here too?

File: chain/app/dbApp.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package app
```
Comment (sorawit) [inline chain/app/dbApp.go]: change file name please

File: chain/app/dbApp.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/bandprotocol/d3n/chain/db"
+)
+
+type dbBandApp struct {
+	*bandApp
+	dbBand *db.BandDB
```
Comment (sorawit) [inline chain/app/dbApp.go]: Let's just name this `db`

File: chain/cmd/bandd/main.go [hunk 4]
```diff
@@ -3,31 +3,37 @@ package main
...
 	dbm "github.com/tendermint/tm-db"
+
+	"github.com/bandprotocol/d3n/chain/app"
+	banddb "github.com/bandprotocol/d3n/chain/db"
 )
 
-const flagInvCheckPeriod = "inv-check-period"
+const (
+	flagInvCheckPeriod = "inv-check-period"
+	flagAddDB          = "add-db"
```
Comment (sorawit) [inline chain/cmd/bandd/main.go]: Let's call this WithDB

File: chain/cmd/bandd/main.go [hunk 5]
```diff
@@ -72,6 +79,21 @@ func main() {
 }
 
 func newApp(logger log.Logger, db dbm.DB, traceStore io.Writer) abci.Application {
+	if viper.IsSet(flagAddDB) {
+		dbSplited := strings.SplitN(viper.GetString(flagAddDB), ":", 2)
```
Comment (sorawit) [inline chain/cmd/bandd/main.go]: Split split split.

File: chain/cmd/bandd/main.go [hunk 6]
```diff
@@ -72,6 +79,21 @@ func main() {
...
+		bandDB, err := banddb.NewDB(dbSplited[0], dbSplited[1])
+		if err != nil {
+			panic(err)
+		}
+		return app.NewDBBandApp(
+			logger, db, traceStore, true, invCheckPeriod, bandDB,
+			baseapp.SetPruning(store.NewPruningOptionsFromString(viper.GetString("pruning"))),
+			baseapp.SetMinGasPrices(viper.GetString(server.FlagMinGasPrices)),
+			baseapp.SetHaltHeight(uint64(viper.GetInt(server.FlagHaltHeight))))
+	}
```
Comment (sorawit) [inline chain/cmd/bandd/main.go:102]: else

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]
Looks good

**Review by sorawit — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**